src/adapter/src/coord/sequencer/inner/copy_from.rs

```diff
@@ -0,0 +1,181 @@
+        // to the Coordinator.
+        let command_tx = self.internal_cmd_tx.clone();
+        let conn_id = ctx.session().conn_id().clone();
+        let closure = Box::new(move |batches| {
+            let _ = command_tx.send(crate::coord::Message::StagedBatches {
+                conn_id,
+                table_id: id,
+                batches,
+            });
+        });
+        // Stash the execute context so we can cancel the COPY.
+        self.active_copies
+            .insert(ctx.session().conn_id().clone(), ctx);
+
+        let _result = self
+            .controller
+            .storage
+            .create_oneshot_ingestion(ingestion_id, collection_id, cluster_id, request, closure)
+            .await;
+    }
+
+    pub(crate) fn commit_staged_batches(
+        &mut self,
+        conn_id: ConnectionId,
+        table_id: CatalogItemId,
+        batches: Vec<Result<ProtoBatch, String>>,
+    ) {
+        let Some(mut ctx) = self.active_copies.remove(&conn_id) else {
+            tracing::warn!(?conn_id, "got response for canceled COPY FROM");
+            return;
+        };
+
+        let mut all_batches = SmallVec::with_capacity(batches.len());
+        let mut all_errors = SmallVec::<[String; 1]>::with_capacity(batches.len());
+        let mut row_count = 0u64;
+
+        for maybe_batch in batches {
+            match maybe_batch {
+                Ok(batch) => {
+                    let count = batch.batch.as_ref().map(|b| b.len).unwrap_or(0);
+                    all_batches.push(batch);
+                    row_count = row_count.saturating_add(count);
+                }
+                Err(err) => all_errors.push(err),
+            }
+        }
+
+        // If we got any errors we need to fail the whole operation.
+        if let Some(error) = all_errors.pop() {
+            tracing::warn!(?error, ?all_errors, "failed COPY FROM");
+
+            // TODO(parkmycar): Cleanup the existing ProtoBatches to prevent leaking them.
+            // TODO(parkmycar): Carry structured errors all the way through.
+
+            ctx.retire(Err(AdapterError::Unstructured(anyhow::anyhow!(
+                "COPY FROM: {error}"
+            ))));
+
+            return;
+        }
+
+        // Stage a WriteOp, then when the Session is retired we complete the
+        // transaction, which handles acquiring the write lock for `table_id`,
+        // advancing the timestamps of the staged batches, and waiting for
+        // everything to complete before sending a response to the client.
+        let stage_write = ctx
+            .session_mut()
+            .add_transaction_ops(TransactionOps::Writes(vec![WriteOp {
+                id: table_id,
+                rows: TableData::Batches(all_batches),
+            }]));
+
+        if let Err(err) = stage_write {
+            ctx.retire(Err(err));
+        } else {
+            ctx.retire(Ok(ExecuteResponse::Copied(row_count.cast_into())));
```

Comment:
Why `Copied` and not `CopyFrom`?

Reply:
The `CopyFrom` execute response is actually what drives the existing `COPY FROM` implementation, so it doesn't really work as the response type here. When ending a session with `ExecuteResponse::CopyFrom` we actually [move the Session to a separate task](https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/protocol.rs#L1740-L1741) which streams in data